Add post on CUDAnative.jl

maleadt opened the pull request:
Finally, some text on CUDAnative.jl! This PR adds 2 blogposts, one for julialang.github.com/blog, the other for NVIDIA's [Parallel Forall blog](https://devblogs.nvidia.com/parallelforall/). Since the NVIDIA blogpost will be detailing Julia's GPU support a bit more in general, I figured it couldn't hurt putting it up here for some feedback as well.

Although both posts share some content, I tried to split the text based on the readers' background:
* **Julia blog**: focus on usage (installation instructions, features & limitations), and how people can help
* **NVIDIA blog**: broader focus on pitching Julia for GPU development, but including some more technical details (eg. comparison of generated assembly, NVIDIA/CUDA-specific technicalities)

cc @vchuravy, @SimonDanisch, @ViralBShah

tkelman commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
```diff
@@ -0,0 +1,234 @@
+[LLVM.jl](https://github.com/maleadt/LLVM.jl) to interface with the framework, and use it to
+compile IR to PTX code, the virtual instruction set used to program CUDA GPUs. Finally, we
+use CUDAdrv.jl to interact with the CUDA driver: upload the PTX code, compile it to
+hardware-specific SASS, and launch the kernel. <!-- TODO: See image 1 for an overview. -->
```
What is SASS? Don't think that acronym is very commonly used

timholy commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
```diff
@@ -0,0 +1,234 @@
+efficient kernel launch sequence, avoiding the runtime overhead typically associated with
+dynamic languages.
+
+<!-- This seems a bit short. Maybe elaborate more? -->
```
Might be worth elaborating on element types here. Could we replace the `Int` with `Float32` or `Float64` and have everything Just Work?

timholy replied:
Also worth putting some kind of link in so that folks who don't already know about `blockIdx` etc can go read about it. Is it worth commenting on why `blockIdx` is a function rather than a constant?

maleadt replied:
This is for the NVIDIA ParallelForall blog, I think all people reading it will be familiar with those functions.

timholy commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
```diff
@@ -0,0 +1,234 @@
+<!-- Mention libNVVM? -->
+
+Behind the scenes, CUDAnative.jl hooks into the Julia compiler to lower kernel functions to
+LLVM IR, an intermediate form of code used by the LLVM compiler framework. We've created
```
might be worth writing "Intermediate Representation" just so the IR acronym is clearer.

timholy commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
```diff
@@ -0,0 +1,234 @@
+Behind the scenes, CUDAnative.jl hooks into the Julia compiler to lower kernel functions to
+LLVM IR, an intermediate form of code used by the LLVM compiler framework. We've created
+[LLVM.jl](https://github.com/maleadt/LLVM.jl) to interface with the framework, and use it to
+compile IR to PTX code, the virtual instruction set used to program CUDA GPUs. Finally, we
```
Maybe worth showing the PTX for this kernel?

maleadt replied:
The `kernel_vadd` PTX isn't terribly interesting, and I already show some for `reduce_warp` where specialization is much more pronounced (don't want to bloat the post with too much assembly).

timholy commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
```diff
@@ -0,0 +1,234 @@
+slightly different semantics. That means you can use all features of the Julia language,
+like metaprogramming or dynamically-typed multimethods, to implement GPU functions. There
+are still some limitations, for example we haven't ported Julia's runtime library or
+implemented a GPU garbage-collector yet, but these aren't structural limitations.
```
Can you be more explicit about the kinds of restrictions? For example, I suspect I can't call a julia function in the middle of a kernel.

vchuravy replied:
Maybe it is enough to link to https://github.com/JuliaGPU/CUDAnative.jl#julia-support and put a more extensive list there?

timholy commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
```diff
@@ -0,0 +1,234 @@
+
+### Future developments
+
+The combination of carefully-designed language semantics and a specializing JIT compiler
```
"Julia's combination"?

timholy commented on blog/_posts/2017-01-24-cudanative.md:
```diff
@@ -0,0 +1,247 @@
+
+### What is missing?
+
+As I've already hinted to, we don't support all features of the Julia language yet. For
```
delete "to"

vchuravy commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
```diff
@@ -0,0 +1,234 @@
+organization](https://github.com/JuliaGPU/). Each of those packages is just a single
+`Pkg.add(...)` away!
+
+What's missing from this list is the lowest abstraction level, where you can create kernels
```
`write kernels` instead of `create`? to emphasis that we are not loading C-kernels.

vchuravy commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
`to parse and compile`

vchuravy commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
`This will allow us to take advantage of the native device libraries, like cuRAND, that NVIDIA provides.` or something to that extend.

vchuravy commented on blog/_posts/2017-01-24-cudanative.md:
```diff
@@ -0,0 +1,247 @@
+
+The programming support we're demonstrating here today consists of the low-level building
+blocks, sitting at the same abstraction level of CUDA C. You should be interested if you
+know (or want to learn) how to program a parallel accelerator like a GPU, while dealing with
```
`, like a GPU, in native Julia, while taking advantage of low-level communication semantics.`

maleadt commented:
Thanks for the feedback, much appreciated! Will work on a new version tomorrow.

A comment from @ViralBShah:
> For many people, this may be their first introduction to julia. Do you think we can use `+(a::CuArray, b::CuArray)` instead of `kernel_vadd`, to show off multiple dispatch and operator overloading, and that the assembly generated is clean despite these high level features? I think it would be a subtle and quiet point.

The problem with kernels is that they can't return values, and hence the addition would need to be in-place. And building the addition on top of `broadcast!` is not supported by `CuArray` (@SimonDanisch is working on this with GPUArrays.jl) and would complicate the simple example, especially for newcomers.

maleadt commented:
Comments addressed. I'll leave this here until I'm happy with the state of CUDAdrv/CUDAnative (there's some recently filed bugs I still need to take a proper look at, and the README's need some work).

ViralBShah commented:
May be good to time it after the release of 0.6-alpha too.

kmsquire commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
```diff
@@ -0,0 +1,233 @@
+By reusing the Julia compiler to generate IR we avoid multiple language implementations with
+slightly different semantics. That means you can use all features of the Julia language,
+like metaprogramming or dynamically-typed multimethods, to implement GPU functions. There
+are still [some limitations]({{page.cudanative_tree}}#julia-support), for example we haven't
```
```
There are still some limitations, for example we haven't ...
```
should probably be either
```
There are still some limitations.  For example, we haven't ...
```
or
```
There are still some limitations--for example, we haven't ...

kmsquire commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
```diff
@@ -0,0 +1,233 @@
+    offset = CUDAnative.warpsize() ÷ 2
+    while offset > 0
+        val = op(val, shfl_down(val, offset))
+        offset ÷= 2
```
Someone reading this who is not familiar with Julia might wonder why you're using `÷` instead of `/`.  Might be worth a brief comment (either inline or below).

kmsquire commented on blog/_posts/2017-01-24-cudanative-nvidia.md:
Link to docs.nvidia.com/cuda/curand/ ?

kmsquire commented on blog/_posts/2017-01-24-cudanative.md:
```diff
@@ -0,0 +1,247 @@
+
+
+
+## ~~Hello World~~ Vector addition
```
Maybe 
```
## Vector Addition (*"Hello World" for the GPU*)
```

kmsquire commented:
Interesting articles--thanks!

ViralBShah commented:
I think now that 0.6 alpha is out, and assuming CUDANative.jl works with it, we should merge these.

maleadt commented:
Yeah, it works on the alpha. I've updated the PR with some minor fixes, and rebased it.
I've also added a piece on the new code reflection (e35dc55b83eac511c109049f5dbf5f0a362a0ead) because that's pretty cool!
Let's merge this after JuliaLang/METADATA.jl#8306 lands.

ViralBShah commented:
Landed.

maleadt commented:
Pushed some final changes, let's publish this.
